rfd/0140-azure-blob-sessions.md

```diff
@@ -0,0 +1,41 @@
+---
+authors: Andrey Bulgakov (andrey.bulgakov@goteleport.com), Edoardo Spadolini (edoardo.spadolini@goteleport.com)
+state: draft
+---
+
+# RFD 0140 - Azure Blob Storage session storage
+
+## Required Approvers
+* Engineering: TBD
+* Security: TBD
+* Product: TBD
+
+## What
+
+Enable users to leverage the Azure Blob Storage managed object storage service for Teleport session recordings.
+
+## Why
+
+To support deployments of Teleport in Azure with the same convenience and reliability that's currently available when using managed storage in AWS and GCP.
+
+## Details
+
+Azure Blob Storage offers similar functionality to S3, including all we need for `lib/events.MultipartUploader` in the strictest sense. In addition to that, it's possible to configure containers (the equivalent of a S3 bucket) to have a time-based immutability policy that prevents moving, renaming, overwriting or deleting a blob for a fixed amount of time after it's first written, which seems to be exactly what we're currently using versioning for.
+
+Similarly to other cloud-based session storage, the configuration looks like this, specifying a storage account URL.
+
+```yaml
+teleport:
+  storage:
+    ...
+
+    audit_sessions_uri: "azblob://<storage account name>.blob.core.windows.net"
+```
+
+It's possible to specify a client ID for a managed identity (to allow for different identities to be used for backend, events and sessions, as opposed to setting the `AZURE_CLIENT_ID` envvar and using the default credentials) by specifying a URL fragment of `#azure_client_id=11111111-2222-3333-4444-555555555555`. The `azblob` schema is only used by Teleport to identify the storage backend, and will actually result in a `https` URL. For testing against simulators (or other services with the same API surface) over `http`, the `azblob-http` schema is also supported (but we should probably leave it undocumented for end users).
+
+Teleport will require access to two containers in the storage account, named `session`, which will store the completed session files, and `inprogress`, which will be used as temporary space to hold parts.
```

Comment:
Open question: do we make the container names configurable, or do we just tell users to create a storage account for Teleport? s3sessions allows using the same bucket with different paths, which is roughly the equivalent (other than having to pick two names instead of just one).